codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/auth/SigV4AsymmetricAuthSchemeIntegration.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.auth
+
+import software.amazon.smithy.aws.traits.auth.SigV4ATrait
+import software.amazon.smithy.aws.traits.auth.UnsignedPayloadTrait
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.codegen.core.SymbolReference
+import software.amazon.smithy.kotlin.codegen.KotlinSettings
+import software.amazon.smithy.kotlin.codegen.core.*
+import software.amazon.smithy.kotlin.codegen.integration.AuthSchemeHandler
+import software.amazon.smithy.kotlin.codegen.integration.KotlinIntegration
+import software.amazon.smithy.kotlin.codegen.model.buildSymbol
+import software.amazon.smithy.kotlin.codegen.model.hasTrait
+import software.amazon.smithy.kotlin.codegen.model.knowledge.AwsSignatureVersion4Asymmetric
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.*
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.knowledge.ServiceIndex
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ShapeId
+
+/**
+ * Register support for the `aws.auth#sigv4a` auth scheme.
+ */
+class SigV4AsymmetricAuthSchemeIntegration : KotlinIntegration {
+    override val order: Byte = -59
+
+    // Needs to be true due to the way integrations are filtered out before application and sigV4a customization.
+    // See 'CodegenVisitor' & 'SigV4AsymmetricTraitCustomization'
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean = true
```

Comment:
**Question**: We don't have any other integrations where `enabledForService` _always_ returns `true`. The majority of them evaluate some criteria from the model and/or codegen context to make the determination about enablement. It's not clear why we need _two_ customizations to handle this. Why couldn't this be combined with `SigV4AsymmetricTraitCustomization` up in **aws-sdk-kotlin**?

Reply:
> We don't have any other integrations where enabledForService always returns true. The majority of them evaluate some criteria from the model and/or codegen context to make the determination about enablement

This one is a weird one, the code in `CodegenVisitor` first filters out the integrations that won't be enabled via the `enabledForService()`  function and AFTER it applies the changes from the integrations.

The issue with this is that we have another integration (`SigV4AsymmetricTraitCustomization`) that applies the SIGV4A trait (the condition we would use to apply this integration). So the integration would be filtered out if we used the `enabledForService()` function because the filtering happens before the customization adds the trait we'd be looking for .

> Why couldn't this be combined with SigV4AsymmetricTraitCustomization up in aws-sdk-kotlin?

The `SigV4AsymmetricTraitCustomization` is there to add the SIGV4A trait to the models that should have it. This one is how we would deal with the trait. 

Since the trait can be added to any model it makes sense to have them be independent, the rendering can be done by smithy and the customization can be done in the sdk only

Reply:
Discussed offline and decided to add a `TODO` + backlog item for improving the filtering logic in `CodegenVisitor` so this `enabledForService` check can be more meaningful.

Reply:
This could be modeled accurately by looking for the `sigv4a` trait and inspecting the endpoint rules.